File: api/formatters.py
from datetime import datetime

def format_timestamp(ts):
    if not ts:
        return None
    try:
        return datetime.utcfromtimestamp(int(ts)).strftime('%Y-%m-%d %H:%M:%S')
    except:
        return ts  # fallback if already formatted

def format_virustotal(data):
    """Enhanced VirusTotal formatting with comprehensive details"""
    if not isinstance(data, dict):
        return {"message": "Invalid VT data format."}
    if "error" in data:
        return data
    if "message" in data:
        return {"message": data["message"]}

    if "details" in data:
        attributes = data["details"].get("data", {}).get("attributes", {})
        vt_id = data["details"].get("data", {}).get("id")
    else:
        attributes = data.get("data", {}).get("attributes", {})
        vt_id = data.get("data", {}).get("id")

    result = {
        "ID": vt_id,
        "Reputation": attributes.get("reputation"),
        "Type": attributes.get("type_description"),
        "Meaningful Name": attributes.get("meaningful_name"),
    }

    # Analysis statistics
    stats = attributes.get("last_analysis_stats", {})
    if stats:
        result["Detection Ratio"] = f"{stats.get('malicious', 0)}/{stats.get('malicious', 0) + stats.get('harmless', 0) + stats.get('undetected', 0) + stats.get('suspicious', 0)}"
        result["Malicious Detections"] = stats.get("malicious", 0)
        result["Suspicious Detections"] = stats.get("suspicious", 0)
        result["Harmless Detections"] = stats.get("harmless", 0)
        result["Undetected"] = stats.get("undetected", 0)
        result["Timeout"] = stats.get("timeout", 0)

    # Categories from different engines
    categories = attributes.get("categories")
    if categories:
        result["Categories"] = categories

    # Timestamps
    if "first_submission_date" in attributes:
        result["First Submission"] = format_timestamp(attributes["first_submission_date"])
    if "last_analysis_date" in attributes:
        result["Last Analysis"] = format_timestamp(attributes["last_analysis_date"])
    if "last_modification_date" in attributes:
        result["Last Modified"] = format_timestamp(attributes["last_modification_date"])

    # Network information
    if "network" in attributes:
        result["Network"] = attributes.get("network")
    if "country" in attributes:
        result["Country"] = attributes.get("country")
    if "asn" in attributes:
        result["ASN"] = attributes.get("asn")
    if "as_owner" in attributes:
        result["AS Owner"] = attributes.get("as_owner")

    # Additional file information
    if "size" in attributes:
        result["File Size"] = f"{attributes['size']:,} bytes"
    if "md5" in attributes:
        result["MD5"] = attributes["md5"]
    if "sha1" in attributes:
        result["SHA1"] = attributes["sha1"]
    if "sha256" in attributes:
        result["SHA256"] = attributes["sha256"]
    if "magic" in attributes:
        result["File Magic"] = attributes["magic"]
    if "signature_info" in attributes:
        sig_info = attributes["signature_info"]
        if isinstance(sig_info, dict):
            result["Signature Subject"] = sig_info.get("subject", "N/A")
            result["Signature Issuer"] = sig_info.get("issuer", "N/A")
            result["Signature Valid"] = sig_info.get("verified", "Unknown")

    # Names and aliases
    names = attributes.get("names", [])
    if names:
        result["Known Names"] = names[:5]  # Limit to first 5

    # Sandbox information
    sandbox_verdicts = attributes.get("sandbox_verdicts", {})
    if sandbox_verdicts:
        sandbox_results = []
        for engine, verdict in sandbox_verdicts.items():
            if isinstance(verdict, dict):
                sandbox_results.append(f"{engine}: {verdict.get('category', 'N/A')}")
        if sandbox_results:
            result["Sandbox Verdicts"] = sandbox_results[:3]  # Top 3

    # Flatten related fields if available
    related = data.get("related", {})

    def extract_ids(entries, limit=5):
        return [e.get("id") for e in entries[:limit] if isinstance(e, dict)]

    if "resolutions" in related and related["resolutions"]:
        result["Resolved IPs"] = extract_ids(related["resolutions"])
    if "communicating_files" in related and related["communicating_files"]:
        result["Communicating Files"] = extract_ids(related["communicating_files"])
    if "downloaded_files" in related and related["downloaded_files"]:
        result["Downloaded Files"] = extract_ids(related["downloaded_files"])
    if "contacted_domains" in related and related["contacted_domains"]:
        result["Contacted Domains"] = extract_ids(related["contacted_domains"])
    if "contacted_ips" in related and related["contacted_ips"]:
        result["Contacted IPs"] = extract_ids(related["contacted_ips"])

    return result

def format_abuseipdb(data):
    """Enhanced AbuseIPDB formatting with complete details"""
    if not isinstance(data, dict):
        return data
    if data.get("error"):
        return data
    
    result = {}
    
    # Core information
    if "IP Address" in data:
        result["IP Address"] = data["IP Address"]
    if "Abuse Score" in data:
        result["Abuse Confidence"] = f"{data['Abuse Score']}%"
        
        # Risk assessment based on score
        score = data["Abuse Score"]
        if score >= 75:
            result["Risk Level"] = "High Risk"
        elif score >= 25:
            result["Risk Level"] = "Medium Risk"
        else:
            result["Risk Level"] = "Low Risk"
    
    # Geographic information
    if "Country" in data:
        result["Country Code"] = data["Country"]
    if "ISP" in data:
        result["ISP"] = data["ISP"]
    if "Domain" in data:
        result["Domain"] = data["Domain"]
    if "Usage Type" in data:
        result["Usage Type"] = data["Usage Type"]
    
    # Abuse information
    if "Total Reports" in data:
        result["Total Reports"] = data["Total Reports"]
    if "Last Reported" in data:
        result["Last Reported"] = data["Last Reported"]
    
    # Additional context
    if "⚠️ Resolved Lookup" in data:
        result["Resolution Note"] = data["⚠️ Resolved Lookup"]
    
    return result

def format_shodan(data):
    """Enhanced Shodan formatting with complete details"""
    if not isinstance(data, dict):
        return data
    if data.get("error"):
        return data
    
    result = {}
    
    # Basic information
    if "ip" in data:
        result["IP Address"] = data["ip"]
    if "organization" in data:
        result["Organization"] = data["organization"]
    if "isp" in data:
        result["ISP"] = data["isp"]
    if "os" in data:
        result["Operating System"] = data["os"]
    
    # Location information
    if "country" in data:
        result["Country"] = data["country"]
    if "city" in data:
        result["City"] = data["city"]
    if "region_code" in data:
        result["Region"] = data["region_code"]
    
    # Network information
    if "open_ports" in data and data["open_ports"]:
        result["Open Ports"] = data["open_ports"]
        result["Total Open Ports"] = len(data["open_ports"])
    if "hostnames" in data and data["hostnames"]:
        result["Hostnames"] = data["hostnames"]
    
    # Vulnerability information
    if "vulns" in data and data["vulns"]:
        result["Vulnerabilities"] = data["vulns"]
        result["Vulnerability Count"] = len(data["vulns"])
    
    # Tags and classification
    if "tags" in data and data["tags"]:
        result["Tags"] = data["tags"]
    
    # Last update
    if "last_update" in data:
        result["Last Updated"] = data["last_update"]
    
    return result

def format_otx(data):
    """Enhanced OTX formatting with complete details"""
    if not isinstance(data, dict):
        return data
    if data.get("error"):
        return data
    
    result = {}
    
    # Basic indicator information
    if "indicator" in data:
        result["Indicator"] = data["indicator"]
    if "type" in data:
        result["Type"] = data["type"]
    if "type_title" in data:
        result["Type Description"] = data["type_title"]
    
    # Pulse information
    pulse_info = data.get("pulse_info", {})
    if pulse_info:
        pulses = pulse_info.get("pulses", [])
        if pulses:
            result["Pulse Count"] = len(pulses)
            
            # Extract key pulse information
            pulse_names = []
            pulse_tags = set()
            malware_families = set()
            
            for pulse in pulses[:5]:  # Limit to first 5 pulses
                if pulse.get("name"):
                    pulse_names.append(pulse["name"])
                if pulse.get("tags"):
                    pulse_tags.update(pulse["tags"])
                if pulse.get("malware_families"):
                    for family in pulse["malware_families"]:
                        if isinstance(family, dict) and family.get("display_name"):
                            malware_families.add(family["display_name"])
            
            if pulse_names:
                result["Related Pulses"] = pulse_names
            if pulse_tags:
                result["Associated Tags"] = list(pulse_tags)[:10]  # Limit tags
            if malware_families:
                result["Malware Families"] = list(malware_families)
    
    # General information
    general = data.get("general", {})
    if general:
        if "sections" in general:
            result["Available Sections"] = general["sections"]
        if "whois" in general:
            result["Whois Available"] = "Yes"
        if "reputation" in general:
            result["Reputation"] = general["reputation"]
    
    return result

def format_threatfox(data):
    """Enhanced ThreatFox formatting with complete details"""
    if not data:
        return {"message": "No ThreatFox data returned."}

    if isinstance(data, dict):
        if data.get("error"):
            return data
        if data.get("status") or data.get("message"):
            return {
                "Status": data.get("status"),
                "Message": data.get("message"),
                "Reason": data.get("reason", "No reason provided")
            }

    if isinstance(data, list) and data:
        formatted = []
        for entry in data:
            formatted_entry = {
                "IOC ID": entry.get("id"),
                "IOC Value": entry.get("ioc"),
                "IOC Type": entry.get("ioc_type"),
                "Threat Type": entry.get("threat_type"),
                "Malware Family": entry.get("malware"),
                "Malware Alias": entry.get("malware_alias"),
            }
            
            # Confidence and reliability
            if entry.get("confidence") is not None:
                formatted_entry["Confidence Level"] = f"{entry.get('confidence')}%"
                confidence = entry.get("confidence")
                if confidence >= 75:
                    formatted_entry["Reliability"] = "High"
                elif confidence >= 50:
                    formatted_entry["Reliability"] = "Medium"
                else:
                    formatted_entry["Reliability"] = "Low"
            
            # Network information
            if entry.get("asn"):
                formatted_entry["ASN"] = entry.get("asn")
            if entry.get("country"):
                formatted_entry["Country"] = entry.get("country")
            
            # Timeline
            if entry.get("first_seen"):
                formatted_entry["First Seen"] = entry.get("first_seen")
            if entry.get("last_seen"):
                formatted_entry["Last_seen"] = entry.get("last_seen")
            
            # Submission details
            if entry.get("uuid"):
                formatted_entry["UUID"] = entry.get("uuid")
            if entry.get("reporter"):
                formatted_entry["Reporter"] = entry.get("reporter")
            if entry.get("credits"):
                formatted_entry["Credits"] = f"{entry.get('credits')} credits"
            
            # Tags and references
            tags = entry.get("tags", [])
            if isinstance(tags, list) and tags:
                formatted_entry["Tags"] = ", ".join(tags)
            elif tags:
                formatted_entry["Tags"] = str(tags)
                
            if entry.get("reference"):
                formatted_entry["Reference"] = entry.get("reference")
            
            formatted.append(formatted_entry)
        
        return formatted

    return {"message": "No matching IOCs found."}

def format_urlhaus(data):
    """Enhanced URLhaus formatting with complete details"""
    if not isinstance(data, dict):
        return data
    if data.get("error"):
        return data
    if data.get("message"):
        return data
    
    result = {}
    
    # Handle search results
    if "results" in data and data["results"]:
        results = []
        for item in data["results"]:
            result_item = {
                "URL": item.get("url"),
                "Host": item.get("host"),
                "Threat Classification": item.get("threat"),
                "Date Added": item.get("date_added"),
                "Reporter": item.get("reporter"),
                "URL Status": item.get("url_status"),
            }
            if item.get("tags"):
                result_item["Tags"] = ", ".join(item["tags"]) if isinstance(item["tags"], list) else item["tags"]
            results.append(result_item)
        return {"results": results, "found": True}
    
    # Handle single payload result
    if data.get("found"):
        if "SHA256" in data:
            result["SHA256 Hash"] = data["SHA256"]
        if "MD5" in data:
            result["MD5 Hash"] = data["MD5"]
        if "File Size" in data:
            result["File Size"] = data["File Size"]
        if "File Type" in data:
            result["File Type"] = data["File Type"]
        if "First Seen" in data:
            result["First Seen"] = data["First Seen"]
        if "Last Seen" in data:
            result["Last Seen"] = data["Last Seen"]
        if "URL Count" in data:
            result["Associated URLs"] = data["URL Count"]
        if "URLs" in data and data["URLs"]:
            result["Sample URLs"] = data["URLs"]
        
        return {**result, "found": True}
    
    return data

def format_malwarebazaar(data):
    """Enhanced MalwareBazaar formatting with complete details"""
    if not isinstance(data, dict):
        return data
    if data.get("error"):
        return data
    if data.get("message"):
        return data
    
    result = {}
    
    # Handle search results
    if "results" in data and data["results"]:
        results = []
        for item in data["results"]:
            result_item = {
                "SHA256 Hash": item.get("SHA256"),
                "File Name": item.get("File Name"),
                "File Type": item.get("File Type"),
                "Malware Signature": item.get("Signature"),
                "File Size": item.get("File Size"),
                "First Seen": item.get("First Seen"),
                "Reporter": item.get("Reporter"),
            }
            if item.get("Tags"):
                result_item["Tags"] = ", ".join(item["Tags"]) if isinstance(item["Tags"], list) else item["Tags"]
            results.append(result_item)
        return {"results": results, "found": True}
    
    # Handle single result
    if data.get("found"):
        if "SHA256" in data:
            result["SHA256 Hash"] = data["SHA256"]
        if "File Name" in data:
            result["File Name"] = data["File Name"]
        if "File Type" in data:
            result["MIME Type"] = data["File Type"]
        if "File Size" in data:
            result["File Size"] = f"{data['File Size']:,} bytes" if isinstance(data["File Size"], int) else data["File Size"]
        if "Signature" in data:
            result["Malware Signature"] = data["Signature"]
        if "Tags" in data:
            tags = data["Tags"]
            result["Tags"] = ", ".join(tags) if isinstance(tags, list) else str(tags)
        if "Delivery Method" in data:
            result["Delivery Method"] = data["Delivery Method"]
        if "First Seen" in data:
            result["First Seen"] = data["First Seen"]
        if "Last Seen" in data:
            result["Last Seen"] = data["Last Seen"]
        if "Reporter" in data:
            result["Reporter"] = data["Reporter"]
        if "Comment" in data and data["Comment"]:
            result["Comment"] = data["Comment"]
        
        # Vendor intelligence
        vendor_intel = data.get("Vendor Detections", {})
        if vendor_intel and isinstance(vendor_intel, dict):
            intel_summary = []
            for vendor, info in vendor_intel.items():
                if isinstance(info, dict):
                    if info.get("verdict") == "MALICIOUS":
                        intel_summary.append(f"{vendor}: {info.get('verdict')}")
                    elif info.get("detection"):
                        intel_summary.append(f"{vendor}: {info.get('detection')}")
            if intel_summary:
                result["Vendor Intelligence"] = intel_summary[:5]  # Limit to 5
        
        return {**result, "found": True}
    
    return data

def format_ipinfo(data):
    """Enhanced IPInfo formatting with complete details"""
    if not isinstance(data, dict):
        return data
    if data.get("error"):
        return data
    
    result = {}
    
    # Basic information
    if "IP" in data:
        result["IP Address"] = data["IP"]
    if "Hostname" in data:
        result["Hostname"] = data["Hostname"]
    
    # Geographic information
    if "City" in data:
        result["City"] = data["City"]
    if "Region" in data:
        result["Region/State"] = data["Region"]
    if "Country" in data:
        result["Country"] = data["Country"]
    
    # Network information
    if "Organization" in data:
        result["Organization"] = data["Organization"]
    if "ASN" in data:
        result["ASN"] = data["ASN"]
    
    # Additional context
    if "⚠️ Resolved Lookup" in data:
        result["Resolution Note"] = data["⚠️ Resolved Lookup"]
    
    return result

def format_urlscan(data):
    """Enhanced URLScan formatting - handled separately in frontend"""
    # URLScan has complex structure handled in frontend
    return data

File: frontend/utils/formatters.js
// utils/formatters.js
// Converted from api/formatters.py

function formatTimestamp(ts) {
    if (!ts) return null;
    try {
      return new Date(parseInt(ts) * 1000).toISOString().replace('T', ' ').substring(0, 19);
    } catch {
      return ts; // fallback if already formatted
    }
  }
  
  export function formatVirusTotal(data) {
    if (!data || typeof data !== 'object') {
      return { message: "Invalid VT data format." };
    }
    if (data.error) return data;
    if (data.message) return { message: data.message };
  
    let attributes, vtId;
    if (data.details) {
      attributes = data.details?.data?.attributes || {};
      vtId = data.details?.data?.id;
    } else {
      attributes = data.data?.attributes || {};
      vtId = data.data?.id;
    }
  
    const result = {
      "ID": vtId,
      "Reputation": attributes.reputation,
      "Type": attributes.type_description,
      "Meaningful Name": attributes.meaningful_name,
    };
  
    // Analysis statistics
    const stats = attributes.last_analysis_stats || {};
    if (Object.keys(stats).length > 0) {
      const total = (stats.malicious || 0) + (stats.harmless || 0) + (stats.undetected || 0) + (stats.suspicious || 0);
      result["Detection Ratio"] = `${stats.malicious || 0}/${total}`;
      result["Malicious Detections"] = stats.malicious || 0;
      result["Suspicious Detections"] = stats.suspicious || 0;
      result["Harmless Detections"] = stats.harmless || 0;
      result["Undetected"] = stats.undetected || 0;
      result["Timeout"] = stats.timeout || 0;
    }
  
    // Categories
    if (attributes.categories) {
      result["Categories"] = attributes.categories;
    }
  
    // Timestamps
    if (attributes.first_submission_date) {
      result["First Submission"] = formatTimestamp(attributes.first_submission_date);
    }
    if (attributes.last_analysis_date) {
      result["Last Analysis"] = formatTimestamp(attributes.last_analysis_date);
    }
    if (attributes.last_modification_date) {
      result["Last Modified"] = formatTimestamp(attributes.last_modification_date);
    }
  
    // Network information
    if (attributes.network) result["Network"] = attributes.network;
    if (attributes.country) result["Country"] = attributes.country;
    if (attributes.asn) result["ASN"] = attributes.asn;
    if (attributes.as_owner) result["AS Owner"] = attributes.as_owner;
  
    // File information
    if (attributes.size) result["File Size"] = `${attributes.size.toLocaleString()} bytes`;
    if (attributes.md5) result["MD5"] = attributes.md5;
    if (attributes.sha1) result["SHA1"] = attributes.sha1;
    if (attributes.sha256) result["SHA256"] = attributes.sha256;
    if (attributes.magic) result["File Magic"] = attributes.magic;
  
    // Signature info
    if (attributes.signature_info) {
      const sigInfo = attributes.signature_info;
      if (typeof sigInfo === 'object') {
        result["Signature Subject"] = sigInfo.subject || "N/A";
        result["Signature Issuer"] = sigInfo.issuer || "N/A";
        result["Signature Valid"] = sigInfo.verified || "Unknown";
      }
    }
  
    // Known names
    const names = attributes.names || [];
    if (names.length > 0) {
      result["Known Names"] = names.slice(0, 5);
    }
  
    // Sandbox verdicts
    const sandboxVerdicts = attributes.sandbox_verdicts || {};
    if (Object.keys(sandboxVerdicts).length > 0) {
      const sandboxResults = [];
      for (const [engine, verdict] of Object.entries(sandboxVerdicts)) {
        if (typeof verdict === 'object') {
          sandboxResults.push(`${engine}: ${verdict.category || 'N/A'}`);
        }
      }
      if (sandboxResults.length > 0) {
        result["Sandbox Verdicts"] = sandboxResults.slice(0, 3);
      }
    }
  
    // Related data
    const related = data.related || {};
    const extractIds = (entries, limit = 5) => {
      return entries.slice(0, limit).map(e => typeof e === 'object' ? e.id : e).filter(Boolean);
    };
  
    if (related.resolutions?.length > 0) {
      result["Resolved IPs"] = extractIds(related.resolutions);
    }
    if (related.communicating_files?.length > 0) {
      result["Communicating Files"] = extractIds(related.communicating_files);
    }
    if (related.downloaded_files?.length > 0) {
      result["Downloaded Files"] = extractIds(related.downloaded_files);
    }
    if (related.contacted_domains?.length > 0) {
      result["Contacted Domains"] = extractIds(related.contacted_domains);
    }
    if (related.contacted_ips?.length > 0) {
      result["Contacted IPs"] = extractIds(related.contacted_ips);
    }
  
    return result;
  }
  
  export function formatAbuseIPDB(data) {
    if (!data || typeof data !== 'object') return data;
    if (data.error) return data;
    
    const result = {};
    
    if (data["IP Address"]) result["IP Address"] = data["IP Address"];
    if (data["Abuse Score"] !== undefined) {
      result["Abuse Confidence"] = `${data["Abuse Score"]}%`;
      
      const score = data["Abuse Score"];
      if (score >= 75) result["Risk Level"] = "High Risk";
      else if (score >= 25) result["Risk Level"] = "Medium Risk";
      else result["Risk Level"] = "Low Risk";
    }
    
    if (data["Country"]) result["Country Code"] = data["Country"];
    if (data["ISP"]) result["ISP"] = data["ISP"];
    if (data["Domain"]) result["Domain"] = data["Domain"];
    if (data["Usage Type"]) result["Usage Type"] = data["Usage Type"];
    if (data["Total Reports"]) result["Total Reports"] = data["Total Reports"];
    if (data["Last Reported"]) result["Last Reported"] = data["Last Reported"];
    if (data["⚠️ Resolved Lookup"]) result["Resolution Note"] = data["⚠️ Resolved Lookup"];
    
    return result;
  }
  
  export function formatShodan(data) {
    if (!data || typeof data !== 'object') return data;
    if (data.error) return data;
    
    const result = {};
    
    if (data.ip) result["IP Address"] = data.ip;
    if (data.organization) result["Organization"] = data.organization;
    if (data.isp) result["ISP"] = data.isp;
    if (data.os) result["Operating System"] = data.os;
    if (data.country) result["Country"] = data.country;
    if (data.city) result["City"] = data.city;
    if (data.region_code) result["Region"] = data.region_code;
    
    if (data.open_ports?.length > 0) {
      result["Open Ports"] = data.open_ports;
      result["Total Open Ports"] = data.open_ports.length;
    }
    if (data.hostnames?.length > 0) {
      result["Hostnames"] = data.hostnames;
    }
    if (data.vulns?.length > 0) {
      result["Vulnerabilities"] = data.vulns;
      result["Vulnerability Count"] = data.vulns.length;
    }
    if (data.tags?.length > 0) {
      result["Tags"] = data.tags;
    }
    if (data.last_update) {
      result["Last Updated"] = data.last_update;
    }
    
    return result;
  }
  
  export function formatOTX(data) {
    if (!data || typeof data !== 'object') return data;
    if (data.error) return data;
    
    const result = {};
    
    if (data.indicator) result["Indicator"] = data.indicator;
    if (data.type) result["Type"] = data.type;
    if (data.type_title) result["Type Description"] = data.type_title;
    
    const pulseInfo = data.pulse_info || {};
    if (pulseInfo.pulses?.length > 0) {
      const pulses = pulseInfo.pulses;
      result["Pulse Count"] = pulses.length;
      
      const pulseNames = [];
      const pulseTags = new Set();
      const malwareFamilies = new Set();
      
      for (const pulse of pulses.slice(0, 5)) {
        if (pulse.name) pulseNames.push(pulse.name);
        if (pulse.tags) pulse.tags.forEach(tag => pulseTags.add(tag));
        if (pulse.malware_families) {
          pulse.malware_families.forEach(family => {
            if (typeof family === 'object' && family.display_name) {
              malwareFamilies.add(family.display_name);
            }
          });
        }
      }
      
      if (pulseNames.length > 0) result["Related Pulses"] = pulseNames;
      if (pulseTags.size > 0) result["Associated Tags"] = Array.from(pulseTags).slice(0, 10);
      if (malwareFamilies.size > 0) result["Malware Families"] = Array.from(malwareFamilies);
    }
    
    const general = data.general || {};
    if (general.sections) result["Available Sections"] = general.sections;
    if (general.whois) result["Whois Available"] = "Yes";
    if (general.reputation) result["Reputation"] = general.reputation;
    
    return result;
  }
  
  export function formatThreatFox(data) {
    if (!data) return { message: "No ThreatFox data returned." };
  
    if (typeof data === 'object' && !Array.isArray(data)) {
      if (data.error) return data;
      if (data.status || data.message) {
        return {
          "Status": data.status,
          "Message": data.message,
          "Reason": data.reason || "No reason provided"
        };
      }
    }
  
    if (Array.isArray(data) && data.length > 0) {
      return data.map(entry => {
        const formatted = {
          "IOC ID": entry.id,
          "IOC Value": entry.ioc,
          "IOC Type": entry.ioc_type,
          "Threat Type": entry.threat_type,
          "Malware Family": entry.malware,
          "Malware Alias": entry.malware_alias,
        };
        
        if (entry.confidence !== undefined) {
          formatted["Confidence Level"] = `${entry.confidence}%`;
          const confidence = entry.confidence;
          if (confidence >= 75) formatted["Reliability"] = "High";
          else if (confidence >= 50) formatted["Reliability"] = "Medium";
          else formatted["Reliability"] = "Low";
        }
        
        if (entry.asn) formatted["ASN"] = entry.asn;
        if (entry.country) formatted["Country"] = entry.country;
        if (entry.first_seen) formatted["First Seen"] = entry.first_seen;
        if (entry.last_seen) formatted["Last Seen"] = entry.last_seen;
        if (entry.uuid) formatted["UUID"] = entry.uuid;
        if (entry.reporter) formatted["Reporter"] = entry.reporter;
        if (entry.credits) formatted["Credits"] = `${entry.credits} credits`;
        
        const tags = entry.tags || [];
        if (Array.isArray(tags) && tags.length > 0) {
          formatted["Tags"] = tags.join(", ");
        } else if (tags) {
          formatted["Tags"] = String(tags);
        }
        
        if (entry.reference) formatted["Reference"] = entry.reference;
        
        return formatted;
      });
    }
  
    return { message: "No matching IOCs found." };
  }
  
  export function formatURLHaus(data) {
    if (!data || typeof data !== 'object') return data;
    if (data.error) return data;
    if (data.message) return data;
    
    const result = {};
    
    // Handle search results
    if (data.results?.length > 0) {
      const results = data.results.map(item => {
        const resultItem = {
          "URL": item.url,
          "Host": item.host,
          "Threat Classification": item.threat,
          "Date Added": item.date_added,
          "Reporter": item.reporter,
          "URL Status": item.url_status,
        };
        if (item.tags) {
          resultItem["Tags"] = Array.isArray(item.tags) ? item.tags.join(", ") : item.tags;
        }
        return resultItem;
      });
      return { results, found: true };
    }
    
    // Handle single payload result
    if (data.found) {
      if (data.SHA256) result["SHA256 Hash"] = data.SHA256;
      if (data.MD5) result["MD5 Hash"] = data.MD5;
      if (data["File Size"]) result["File Size"] = data["File Size"];
      if (data["File Type"]) result["File Type"] = data["File Type"];
      if (data["First Seen"]) result["First Seen"] = data["First Seen"];
      if (data["Last Seen"]) result["Last Seen"] = data["Last Seen"];
      if (data["URL Count"]) result["Associated URLs"] = data["URL Count"];
      if (data.URLs?.length > 0) result["Sample URLs"] = data.URLs;
      
      return { ...result, found: true };
    }
    
    return data;
  }
  
  export function formatMalwareBazaar(data) {
    if (!data || typeof data !== 'object') return data;
    if (data.error) return data;
    if (data.message) return data;
    
    const result = {};
    
    // Handle search results
    if (data.results?.length > 0) {
      const results = data.results.map(item => {
        const resultItem = {
          "SHA256 Hash": item.SHA256,
          "File Name": item["File Name"],
          "File Type": item["File Type"],
          "Malware Signature": item.Signature,
          "File Size": item["File Size"],
          "First Seen": item["First Seen"],
          "Reporter": item.Reporter,
        };
        if (item.Tags) {
          resultItem["Tags"] = Array.isArray(item.Tags) ? item.Tags.join(", ") : item.Tags;
        }
        return resultItem;
      });
      return { results, found: true };
    }
    
    // Handle single result
    if (data.found) {
      if (data.SHA256) result["SHA256 Hash"] = data.SHA256;
      if (data["File Name"]) result["File Name"] = data["File Name"];
      if (data["File Type"]) result["MIME Type"] = data["File Type"];
      if (data["File Size"]) {
        result["File Size"] = typeof data["File Size"] === 'number' 
          ? `${data["File Size"].toLocaleString()} bytes` 
          : data["File Size"];
      }
      if (data.Signature) result["Malware Signature"] = data.Signature;
      if (data.Tags) {
        const tags = data.Tags;
        result["Tags"] = Array.isArray(tags) ? tags.join(", ") : String(tags);
      }
      if (data["Delivery Method"]) result["Delivery Method"] = data["Delivery Method"];
      if (data["First Seen"]) result["First Seen"] = data["First Seen"];
      if (data["Last Seen"]) result["Last Seen"] = data["Last Seen"];
      if (data.Reporter) result["Reporter"] = data.Reporter;
      if (data.Comment) result["Comment"] = data.Comment;
      
      // Vendor intelligence
      const vendorIntel = data["Vendor Detections"] || {};
      if (vendorIntel && typeof vendorIntel === 'object') {
        const intelSummary = [];
        for (const [vendor, info] of Object.entries(vendorIntel)) {
          if (typeof info === 'object') {
            if (info.verdict === "MALICIOUS") {
              intelSummary.push(`${vendor}: ${info.verdict}`);
            } else if (info.detection) {
              intelSummary.push(`${vendor}: ${info.detection}`);
            }
          }
        }
        if (intelSummary.length > 0) {
          result["Vendor Intelligence"] = intelSummary.slice(0, 5);
        }
      }
      
      return { ...result, found: true };
    }
    
    return data;
  }
  
  export function formatIPInfo(data) {
    if (!data || typeof data !== 'object') return data;
    if (data.error) return data;
    
    const result = {};
    
    if (data.IP) result["IP Address"] = data.IP;
    if (data.Hostname) result["Hostname"] = data.Hostname;
    if (data.City) result["City"] = data.City;
    if (data.Region) result["Region/State"] = data.Region;
    if (data.Country) result["Country"] = data.Country;
    if (data.Organization) result["Organization"] = data.Organization;
    if (data.ASN) result["ASN"] = data.ASN;
    if (data["⚠️ Resolved Lookup"]) result["Resolution Note"] = data["⚠️ Resolved Lookup"];
    
    return result;
  }
  
  export function formatURLScan(data) {
    // URLScan has complex structure handled in frontend
    return data;
  }